### ISSUE TITLE
PYTHONIOENCODING=undefined doesn't work in Python 3

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/cpython#76637]
BPO | [32456](https://bugs.python.org/issue32456)
--- | :---
Nosy | @malemburg, @ncoghlan, @vstinner, @benjaminp, @ezio-melotti, @vadmium, @serhiy-storchaka

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-12-30.16:26:14.027>
labels = ['interpreter-core', 'type-bug', '3.7', 'expert-unicode', 'expert-IO']
title = "PYTHONIOENCODING=undefined doesn't work in Python 3"
updated_at = <Date 2018-01-03.11:01:14.784>
user = 'https://github.com/serhiy-storchaka'
```

bugs.python.org fields:
```python
activity = <Date 2018-01-03.11:01:14.784>
actor = 'serhiy.storchaka'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core', 'Unicode', 'IO']
creation = <Date 2017-12-30.16:26:14.027>
creator = 'serhiy.storchaka'
dependencies = []
files = []
hgrepos = []
issue_num = 32456
keywords = []
message_count = 5.0
messages = ['309237', '309390', '309404', '309406', '309408']
nosy_count = 7.0
nosy_names = ['lemburg', 'ncoghlan', 'vstinner', 'benjamin.peterson', 'ezio.melotti', 'martin.panter', 'serhiy.storchaka']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue32456'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

### Linked PRs
* gh-113834
* gh-136531
* gh-136536
* gh-136537

**Comment by ronaldoussoren** [comment]
For 3.11 and later you get a slightly better behaviour:

```
$ PYTHONIOENCODING=undefined python3.11 -c 'false'
object address  : 0x100dccee0
object refcount : 2
object type     : 0x10229d7a8
object type name: NameError
object repr     : NameError("name 'false' is not defined")
lost sys.stderr
```

That's probably as good as it gets given that the encoding exists, but errors out when trying to use it (by design). 

Non existing encodings give an error during startup that is clear enough to debug the issue:

```
PYTHONIOENCODING=undefined2 python3.11 -c 'false'
Fatal Python error: init_stdio_encoding: failed to get the Python codec name of the stdio encoding
Python runtime state: core initialized
LookupError: unknown encoding: undefined2
```

**Comment by malemburg** [comment]
FYI: The "undefined" codec is meant for testing purposes in Python 3 and was used in Python 2 to prevent auto-coercion of strings to Unicode. It always errors out and will not do any encoding / decoding.

When using it for PYTHONIOENCODING, you essentially prevent any output encoding from happening in Python 3, so it's not really surprising to not get any output or errors from Python 3 in such a case. Specifically for sys.stderr it would make sense to use UTF-8 as fallback in case output to sys.stderr raises an exception, but I'm not sure whether that's possible (if sys.stderr is simply set up as an io stream, there's no special code which could implement such a fallback).

I'd close this as "won't fix", because it's rather unlikely to be used in production situations in Python 3.

**Comment by vstinner** [comment]
The "undefined" codec was added to Python 2.0 (in 2000) by commit 54480d300abcc5b600a5af933b977e775336a5ed:

```
commit 54480d300abcc5b600a5af933b977e775336a5ed
Author: Marc-André Lemburg <mal@egenix.com>
Date:   Wed Jun 7 09:04:05 2000 +0000

    New codec which always raises an exception when used. This
    codec can be used to effectively switch off string coercion
    to Unicode.
```

Example:

```
$ python2.7
Python 2.7.18 (default, Nov 27 2023, 00:00:00) 
>>> import imp, sys
>>> imp.reload(sys)
<module 'sys' (built-in)>

>>> sys.setdefaultencoding('undefined')
>>> sys.getdefaultencoding()
'undefined'

>>> "bytes" + u"unicode"
UnicodeError: undefined encoding
```

There is **dead code** using it in ``site.py``:

```
def setencoding():
    """Set the string encoding used by the Unicode implementation.  The
    default is 'ascii', but if you're willing to experiment, you can
    change this."""
    encoding = "ascii" # Default value set by _PyUnicode_Init()
    if 0:
        # Enable to support locale aware default string encodings.
        import locale
        loc = locale.getdefaultlocale()
        if loc[1]:
            encoding = loc[1]
    if 0:
        # Enable to switch off string to Unicode coercion and implicit
        # Unicode to string conversion.
        encoding = "undefined"
    if encoding != "ascii":
        # On Non-Unicode builds this will raise an AttributeError...
        sys.setdefaultencoding(encoding) # Needs Python Unicode build !
```

I checked the Git history: this code path was always dead code (hidden by `if 0:`).

Note: the site module deletes the `sys.setdefaultencoding()` function:

```py
    # Remove sys.setdefaultencoding() so that users cannot change the
    # encoding after initialization.  The test for presence is needed when
    # this module is run as a script, because this code is executed twice.
    if hasattr(sys, "setdefaultencoding"):
        del sys.setdefaultencoding
```

In Python 3 (Python 3.0 was released in 2008), this codec is no longer needed to prevent coercion:

```
$ python3
Python 3.12.1 (main, Dec 18 2023, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)] on linux
>>> b"bytes" + "unicode"
TypeError: can't concat str to bytes
```

Moreover, `-b` command line option of Python 3 can be used to emit a `BytesWarning` when comparing bytes strings to Unicode strings (use `-bb` to raise an exception instead):

```
$ python3 -b
Python 3.12.1 (main, Dec 18 2023, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)] on linux
>>> b"bytes" == "unicode"
<stdin>:1: BytesWarning: Comparison between bytes and string
False
```

I don't think that this "undefined" codec is still relevant in Python 3 and so I propose starting to deprecating it.

It can be surprising that a codec called "undefined" is actually "defined" :-)

**Comment by vstinner** [comment]
> I don't think that this "undefined" codec is still relevant in Python 3 and so I propose starting to deprecating it.

I wrote PR #113834 for that.

**Comment by vstinner** [comment]
> I guess there was an overlap in our comments.

Right, we basically posted our comments "at the same time" :-) You replied while I was digging Git history.

> It is still relevant for testing purposes in Python 3, since you will want to know what happens to your application in case an encoding setting causes errors, so -1 on deprecating it.

In the Python stdlib, a single function uses the "undefined" codec: it's a regression test on the "undefined" codec. I don't understand what is this "for testing purposes" use case in 3rd party code. Why would anyone want a codec which always fail? If it's only for testing purpose, Lib/encodings/undefined.py can be copied in the projects using it.

For me, it's surprising that this "undefined" name gives a codec in Python.

> Regarding the "dead code" you found (...). This is indeed no longer relevant and can be removed.

It was removed long time ago in Python 3 site.py module.

**Comment by malemburg** [comment]
On 16.01.2024 14:52, Victor Stinner wrote:
> In the Python stdlib, a single function uses the "undefined" codec: 
> it's a regression test on the "undefined" codec. I don't understand 
> what is this "for testing purposes" use case in 3rd party code. 
If you want to check how code behaves when it runs into encoding errors, 
the "undefined" codec provides an easy way to do this, without having to 
construct artificial encoding errors.
>  Why would anyone want a codec which always fail? If it's only for 
> testing purpose, Lib/encodings/undefined.py can be copied in the 
> projects using it. 
No, please keep it. Thanks.

-- 
Marc-Andre Lemburg
https://www.malemburg.com/

**Comment by vstinner** [comment]
I closed my PR https://github.com/python/cpython/pull/113834 which deprecates the "undefined" codec.

**Comment by malemburg** [comment]
On 19.01.2024 10:16, Victor Stinner wrote:
> I closed my PR #113834 <https://github.com/python/cpython/pull/113834> 
> which deprecates the "undefined" codec.

Thank you.

-- 
Marc-Andre Lemburg
https://www.malemburg.com/

**Comment by StanFromIreland** [comment]
If we are keeping the codec only for testing purposes, I propose it is noted as such in the docs. This would fall under [soft-deprecation](https://peps.python.org/pep-0387/#soft-deprecation) from my understanding.

Then this issue can be closed, as it is not expected/worth to implement a workaround for a "testing" encoding, and the current message has improved further from [2024](https://github.com/python/cpython/issues/76637#issuecomment-1881827943), too.

<details>

<summary>Error message</summary>

```
object address  : 0x7ff9bb146fe0
object refcount : 3
object type     : 0x9004e0
object type name: UnicodeError
object repr     : UnicodeError('undefined encoding')
lost sys.stderr
```

</details>

**Comment by malemburg** [comment]
We should document it, but there's no need for a soft deprecation. As I mentioned, it can be used in (coded) tests.
